src/int/specialized_div_rem/mod.rs

```diff
@@ -0,0 +1,314 @@
+//! `inline` so this can happen even with no LTO). For example, most target compilations
+//! do the following:
+//!
+//!  - Many 128 bit division functions like `u128::wrapping_div` use
+//!    `std::intrinsics::unchecked_div`, which gets linked to `__udivti3` because there is
+//!    not a 128 bit by 128 bit hardware division function in most architectures.
+//!    `__udivti3` uses `u128_div_rem` (this extra level of function calls exists because
+//!    `__umodti3` and `__udivmodti4` also exist, and `specialized_div_rem` supplies just
+//!    one function to calculate both the quotient and remainder. The legacy signatures
+//!    with the `rem: Option<&mut ...>` should be optimized away). The `impl_trifecta!`
+//!    defines `u128_div_rem` to use the trifecta algorithm, which requires the half sized
+//!    division `u64_by_u64_div_rem`. If the architecture supplies a 64 bit hardware
+//!    division instruction, linkers and compilers will reduce `u64_by_u64_div_rem` down
+//!    to those instructions. Note that we do not specify the half size division directly
+//!    to be `__udivdi3`, because hardware division would never be introduced.
+//!  - If the architecture does not supply a 64 bit hardware division instruction, linkers
+//!    will look for functions such as `__udivdi3`. This will call `u64_div_rem` which is
+//!    defined by `impl_delegate!` on most architectures. The half division for this
+//!    algorithm is `u32_by_u32_div_rem` which in turn becomes hardware division
+//!    instructions or more software division algorithms.
+//!  - If the architecture does not supply a 32 bit hardware instruction, linkers will
+//!    look for `__udivsi3`. `impl_binary_long!` is used, but this  algorithm uses no half
+//!    division, so the chain of calls ends here.
+//!
+//! On some architectures like x86_64, an asymmetrically sized division is supplied, in
+//! which 128 bit numbers can be divided by 64 bit numbers. `impl_asymmetric!` is used to
+//! extend the 128 by 64 bit division to a full 128 by 128 bit division.
+
+// `allow(dead_code)` is used in various places, because the configuration code would otherwise be
+// ridiculously complex
+
+#[macro_use]
+mod norm_shift;
+
+#[macro_use]
+mod binary_long;
+
+#[macro_use]
+mod delegate;
+
+#[macro_use]
+mod trifecta;
+
+#[macro_use]
+mod asymmetric;
+
+#[inline]
+fn zero_div_fn() -> ! {
+    ::abort()
+}
+
+// TODO: if Rust adds a way to check for the "B" extension on RISC-V, then modify this configuration
+// code and the 128 bit division definitions below. When `USE_LZ` is set to false, the normalization
+// shift calculation uses an algorithm that better tolerates architectures without a fast way to
+// calculate `leading_zeros`. `trifecta` absolutely requires fast `leading_zeros` calculation, so
+// we also change to using the `delegate` algorithm for 128 bits in the case of RISC-V.
+
+#[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
```

Comment:
```suggestion
#[cfg(any(target_arch = "riscv32", target_arch = "riscv64", all(target_arch = "arm", not(target_feature = "v5")))]
```

ARMv4 doesn't have a CLZ instruction, it was added in ARMv5.

Reply:
According to the wikipedia article, it is specifically ARMv5T and later. I'm confused about how configuration with ARM targets work. There are `armv5te-...` target triples. I'm especially confused about what versions the `arm-...` triples encompass. Does `not(target_feature = "v5")` return true on all ARM targets v5 and after?

Reply:
`target_feature = "v5"` does not seem to exist in `rustc --print target-features`. I will keep the current configuration until `target_feature = "lzcnt"` is fixed